Bug 1722455 - Update pings documentation with Glean.js information (#1744)
mozilla/glean: docs/user/user/pings/index.md

- The [`ping_info` section](#the-ping_info-section) contains core metadata
  that is included in **every** ping.

- The [`client_info` section](#the-client_info-section) contains information that identifies the client.
  It is included in most pings (including all built-in pings), but may be excluded from pings
  where we don't want to connect client information with the other metrics in the ping.

The following keys are only present if any metrics or events were recorded for the given ping:

- The `metrics` section contains the submitted values for all metric types
  except for [events](../metrics/event.md). It has keys for each of the metric types,
  under which is data for each metric.

- The `events` section contains the events recorded in the ping.

See the [payload documentation](../../../dev/core/internal/payload.md)
for more details for each metric type in the `metrics` and `events` section.

### The `ping_info` section

The following fields are included in the `ping_info` section, for every ping.
Optional fields are marked accordingly.

#### `seq`

_Type: [Counter](../../reference/metrics/counter.md),
Lifetime: [User](../../reference/yaml/metrics.md#user)_

> A running counter of the number of times pings of this type have been sent

Reviewer: ```suggestion
A running counter of the number of times pings of this type have been sent.
```

```

### The `client_info` section

The following fields are included in the `client_info` section.
Optional fields are marked accordingly.

#### `app_build`

_Type: [String](../../reference/metrics/counter.md),
Lifetime: [Ping](../../reference/yaml/metrics.md#ping-default)_

The build identifier generated by the CI system (e.g. "1234/A"). For language bindings that provide
automatic detection for this value, (e.g. Android/Kotlin), in the unlikely event that
the build identifier can not be retrieved from the OS, it is set to `inaccessible`.
For other language bindings, if the value was not provided through configuration,

> this metric gets set to `Unknown`.

Reviewer: huh, I guess we should have standardized this better...

Reviewer: (for some it's lowercase `unknown`. I might file a bug so we make it the same value everywhere, don't bother about this in the docs though)

Author: Nice catch :)

or a git hash) is application-specific. In the unlikely event this value can not be obtained from the OS,
 it is set to "inaccessible". If it is accessible, but not set by the application, it is set to "Unknown".

#### `architecture`

_Type: [String](../../reference/metrics/counter.md),
Lifetime: [Ping](../../reference/yaml/metrics.md#ping-default)_

The architecture of the device (e.g. "arm", "x86").

#### `client_id` _(optional)_

_Type: [String](../../reference/metrics/counter.md),
Lifetime: [User](../../reference/yaml/metrics.md#user)_

A UUID identifying a profile and allowing user-oriented correlation of data.

#### `device_manufacturer` _(optional)_

_Type: [String](../../reference/metrics/counter.md),
Lifetime: [Application](../../reference/yaml/metrics.md#application)_

The model name of the device. On Android, this is [`Build.MODEL`], the user-visible name of the device.

#### `device_model` _(optional)_

_Type: [String](../../reference/metrics/counter.md),
Lifetime: [Application](../../reference/yaml/metrics.md#application)_

The model of the device the application is running on.

> On Android, this is Build.MODEL, the user-visible marketing name, like "Pixel 2 XL".

Reviewer: ```suggestion
On Android, this is `Build.MODEL`, the user-visible marketing name, like "Pixel 2 XL".
```

Not set if the device model can't be determined (e.g. on Desktop).

#### `first_run_date`

_Type: [Datetime](../../reference/metrics/datetime.md),
Lifetime: [User](../../reference/yaml/metrics.md#user)_

The date of the first run of the application, in local time and with day precision,
including timezone information.

#### `os`

_Type: [String](../../reference/metrics/counter.md),
Lifetime: [Application](../../reference/yaml/metrics.md#application)_

> The name of the operating system (e.g. "linux", "Android", "ios").

Reviewer: ```suggestion
The name of the operating system (e.g. "Linux", "Android", "iOS").
```

To keep resource usage in check, the Glean SDK enforces some limitations on ping uploading and ping storage.

#### Rate limiting

Only up to 10 ping submissions every 60 seconds are allowed. There are no exposed methods
to change these rate limiting defaults, follow [Bug 1647630](https://bugzilla.mozilla.org/show_bug.cgi?id=1647630)
for updates.

Rate limiting is not implemented on the Glean JavaScript SDK yet,
follow [Bug 1712920](https://bugzilla.mozilla.org/show_bug.cgi?id=1712920) for updates.

#### Request body size limiting

The body of a ping request may have up to 1MB (after compression). Pings that exceed this size are discarded
and don't get uploaded. Size and number of discarded pings are recorded on the internal
Glean metric [`glean.upload.discarded_exceeding_pings_size`](../collected-metrics/metrics.md#metrics-1).

#### Storage quota

Pending pings are stored on disk. Storage is scanned every time Glean is initialized and upon scanning

> Glean checks its size. If it exceeds a size of 10MB or 250 pending ping, pings are deleted to get the

Reviewer: ```suggestion
Glean checks its size. If it exceeds a size of 10MB or 250 pending pings, pings are deleted to get the
```

storage back to an accepted size. Pings are deleted oldest first, until the storage size is below the quota.

The number of deleted pings due to exceeding storage quota is recorded on the metric
[`glean.upload.deleted_pings_after_quota_hit`](../collected-metrics/metrics.md#metrics-1)
and the size of the pending pings directory is recorded (regardless on whether quota has been reached)
on the metric [`glean.upload.pending_pings_directory_size`](../collected-metrics/metrics.md#metrics-1).

Deletion request pings are not subject to this limitation and never get deleted.

### Submitted headers

A pre-defined set of headers is additionally sent along with the submitted ping.

#### `Content-Type`

Describes the data sent to the server. Value is always `application/json; charset=utf-8`.

#### `Date`

> Submission date/time in GMT/UTC+0 offset e.g. `Mon, 23 Jan 2019 10:10:10 GMT+00:00`.

Reviewer: ```suggestion
Submission date/time in GMT/UTC+0 offset, e.g. `Mon, 23 Jan 2019 10:10:10 GMT+00:00`.
```

#### `User-Agent`

The Glean SDKs do not set this header[^1], so it will contain whatever value was set
by the underlying uploading mechanism. For example, when sending pings from browsers
it will contain the characteristic [browser UA string](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent#syntax).

> This header is parsed by the Glean pipeline and can be queried at analysis time through

Reviewer: Might be good to call out that in old (actually current) Glean versions this is the same as `X-Telemetry-Agent` in newer ones (at least for Glean core)

Author: I did call it out in the footnote https://github.com/mozilla/glean/pull/1744/files#diff-7941a3cb068acb37c109d8f78146a0b31b52b984271ed3767130a61a0fccb1f8R312-R315

Do you think that is too hidden?

Reviewer: Should be fine when rendered then.

the [`metadata.user_agent.*`](https://docs.telemetry.mozilla.org/datasets/pings.html?highlight=metadata#ping-metadata)
fields in the ping tables.

#### `X-Telemetry-Agent`

> This header is useful for debugging purposes when pings are sent to [the error stream](https://docs.telemetry.mozilla.org/concepts/pipeline/filtering.html?highlight=error%20stream#querying-the-error-stream),

Reviewer: I would start this sentence with the latter part, explaining what it is.
E.g. 

```suggestion
The Glean SDK version and platform this ping is sent from.
Useful for debugging purposes when pings are sent to [the error stream](https://docs.telemetry.mozilla.org/concepts/pipeline/filtering.html?highlight=error%20stream#querying-the-error-stream).
```

as it describes the application and the Glean SDK used for sending the ping.

> It's looks like `Glean/0.40.0 (Kotlin on Android)`, where `0.40.0` is the Glean SDK version number

Reviewer: ```suggestion
It's looks like `Glean/40.0.0 (Kotlin on Android)`, where `40.0.0` is the Glean SDK version number
```

:P

and `Kotlin on Android` is the name of the language used by the SDK that sent the request
plus the name of the platform it is running on.

_This header is currently only sent by the Glean JavaScript SDK. See [note](#1)._

#### `X-Client-Type`

Custom header to support handling of Glean pings in the legacy pipeline. Values is always `Glean`.

#### `X-Client-Version`

The Glean SDK version e.g. `0.40.0`, sent as a custom header to support handling of Glean pings in the legacy pipeline.

#### `X-Debug-Id` _(optional)_

> Debug header attached to Glean pings by using the [debug APIs](../../reference/debug/debugViewTag.md)
> e.g. `test-tag`.

Reviewer: ```suggestion
Debug header attached to Glean pings by using the [debug APIs](../../reference/debug/debugViewTag.md),
e.g. `test-tag`.
```

When this header is present, the ping is redirected to the

> [Glean Debug View](http://localhost:3000/user/debugging/index.html#glean-debug-view).

Reviewer: wrong url.

```suggestion
[Glean Debug View](../../debugging/index.md#glean-debug-view).
```

(double check if this is correct please)

Author: ```suggestion
[Glean Debug View](../debugging/index.md#glean-debug-view).
```

Actually only one level down.
